### ISSUE TITLE
Delve should not attempt to follow procfs exe symlink

### ISSUE THREAD

**Issue by igorwwwwwwwwwwwwwwwwwwww** [go-delve/delve#2168]
1. What version of Delve are you using (`dlv version`)? 1.5.0 (build `7555d1c06339e51a4f1e2d7a62dc9c44745e10b7`)
2. What version of Go are you using? (`go version`)? go1.14.5
3. What operating system and processor architecture are you using? Linux on 64-bit x86.
4. What did you do? Attempt to attach to an existing process whose binary had been replaced via `sudo ./dlv attach 20447 ./gitaly-ssh`.
5. What did you expect to see? A successful attachment.
6. What did you see instead?

```
$ sudo ./dlv attach 20447 ./gitaly-ssh

could not attach to pid 20447: lstat /opt/gitlab/embedded/bin/gitaly-ssh (deleted): no such file or directory
```

Here's the deal: procfs is weird.

When you `stat /proc/<pid>/exe`, it looks like a symlink. Except when the inode of the binary changed since the process started, it appends ` (deleted)` which then is no longer a symlink you can actually follow.

Delve is trying to be nice in following the symlink.

Except `/proc/<pid>/exe` in fact _doesn't_ behave like a symlink when you `open` it. In that case it will give you a fd to the memory mapped (older) copy of the program.

I believe delve should not try to resolve the symlink and instead directly attempt to open the procfs exe here.

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
Here's the line:

https://github.com/go-delve/delve/blob/7555d1c06339e51a4f1e2d7a62dc9c44745e10b7/pkg/proc/native/proc_linux.go#L268-L271

I believe that can simply be:

```
return fmt.Sprintf("/proc/%d/exe", pid) 
```

**Comment by chainhelen** [comment]
#2024

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
Thanks for the links to previous issues!

It certainly is true that in most cases a process will be restarted after its binary was updated. However, there are still many cases in practice where that assumption does not hold.

Given that delve may be used as a last resort to debug hung processes, it seems like a likely case for a delve user to encounter.

My specific case is a process that has been hanging for 2 days, and the binary was replaced in the mean time by continuous deployment. This use case is currently unsupported by delve, and I had to resort to using gdb.

I see that the change to use `EvalSymlinks` was made in order to support a separate code path for finding symbols. My suggestion would be to decouple the two and handle the case of "deleted" executables more gracefully in `openSeparateDebugInfo`.

I believe this is essentially what @aarzilli proposed in https://github.com/go-delve/delve/pull/2025#pullrequestreview-399849902.
